Implement Background Sync and Local Storage for Calendar Event Queries

VanWardawg opened the pull request:
Main change moves from the MS365CalendarData class to the M365CalendarEventSyncManager and the M365CalendarSyncCoordinator which implement the DataUpdateCoordinator home assistant class.
This add the LocalCalendarStore which stores event information in the .storage file of home assistant config.
Events are loaded into a timeline which is then quickly loaded.

All code is based on Gcal's implementation here:
https://github.com/allenporter/gcal_sync/blob/main/gcal_sync/sync.py#L47
and here:
https://github.com/home-assistant/core/blob/dev/homeassistant/components/google/coordinator.py

This has increase my own local speed (Home assistant green) when filtering and using the calendar.

Improvements to consider:
Add tests
M365CalendarEventSyncManager.Run should merge new data with loaded data instead of overwriting (had issues syncing deleted items)
Add configuration of storage time (days +- to sync) as well as data sync frequency in integration configuration
improve data serialization (I believe a bunch is still missing from my local experiment at least)

## Summary by Sourcery

Implements background sync and local storage for calendar events, improving the speed of filtering and using the calendar, and migrates from MS365CalendarData to M365CalendarEventSyncManager and M365CalendarSyncCoordinator.

Enhancements:
- Uses DataUpdateCoordinator for calendar data synchronization.
- Improves calendar event loading speed by storing event information in the Home Assistant .storage file.
- Adds M365CalendarService to manage calendar API calls.
- Adds M365Timeline to manage calendar events.
- Adds M365CalendarEventStoreService to perform event lookups from the local store.
- Adds M365CalendarEventSyncManager to manage synchronizing events from API to local store.
- Adds LocalCalendarStore for local persistence of calendar and event data.

RogerSelwyn commented:
@VanWardawg thank you very much for the PR. It may take a few days to validate and merge. So comments/questions/below.

> Add tests
I should be able to do this
> M365CalendarEventSyncManager.Run should merge new data with loaded data instead of overwriting (had issues syncing deleted items)
Can you expand on the issues you had?
> Add configuration of storage time (days +- to sync) 
Not quite sure what you mean by this. Is it the range of data to pull? Or how long to store for?
> as well as data sync frequency in integration configuration
Do you mean setting this as a configureable setting? If so, I can do that pretty easily and provide into the coordinator setup.
> improve data serialization (I believe a bunch is still missing from my local experiment at least)
What improvements are you looking for, and in particular what do you think is missing? Bear in mind not all data is pulled from the api, only those items configured in the integration.

RogerSelwyn commented:
@VanWardawg just to add, after a minor tidy up I am left with three major linting errors. The old data class is still referenced in the following methods which will cause a failure at run time:

- _async_get_events_and_store - this method is no longer referenced, I think it is no longer needed
- async_create_calendar_event  - the MS365 Calendar (as opposed to HA core) event creation service will fail
- _async_send_response - sending a response to a calendar invitation will fail. This is an MS365 service, there is no HA core equivalent

Can you take a look at these?

VanWardawg commented:
Yes so first thanks for finding and fixing some items! you are much more knowledgeable about the workings of the integration.
To answer your questions:

"Can you expand on the issues you had?"
When new data gets pulled in the merge I was doing, it was not removing the deleted items (from other devices) that were no longer returned from the API. This wasn't bad as I have it syncing the whole time range and do manual queries for any api call outside of the range but if/when this sync wants to slowly add/sync information it would need to be able to merge/add items while removing deleted items.

Is it the range of data to pull? Or how long to store for?
Yes, right now I have the min/max time to sync hardcoded in the coordinator but my thoughts would be having this as a configureable setting
Same goes for the data sync frequency, I have it configured for every 7 minutes but if people aren't changing or don't mind slowly sync times between devices (for less powered devices, etc.) then they could adjust it.

Current values I mean for this are:
SYNC_EVENT_MIN_TIME = timedelta(days=-60)
SYNC_EVENT_MAX_TIME = timedelta(days=90)
and 
MIN_TIME_BETWEEN_UPDATES = timedelta(minutes=7)

What improvements are you looking for, and in particular what do you think is missing?
Everything seems to work fine which is strange to me, I just looked at the storage files the localcalendarstore creates and the serialized events don't seem to have as much details as I was expecting, but it works so...
for example here is a serialized event I have in the storage file (I haven't removed anything for privacy, this is all it shows)
{
            "AQMkADAwATMwMAItYjYyOS0zNTc5LTAwAi0wMAoARgAAAwFauBQmPy1Eif70yCxgWX8HAPQ6RsMlfYhHrAYaya5uDJYAB8Qu_PUAAAD0OkbDJX2IR6wGGsmubgyWAAfHSmVWAAAA": {
              "calendar_id": null,
              "object_id": "AQMkADAwATMwMAItYjYyOS0zNTc5LTAwAi0wMAoARgAAAwFauBQmPy1Eif70yCxgWX8HAPQ6RsMlfYhHrAYaya5uDJYAB8Qu_PUAAAD0OkbDJX2IR6wGGsmubgyWAAfHSmVWAAAA",
              "body_type": "html",
              "has_attachments": false,
              "ical_uid": null,
              "is_cancelled": false,
              "is_organizer": true,
              "locations": [],
              "online_meeting_url": null,
              "online_meeting": null,
              "series_master_id": null,
              "web_link": null
            },

VanWardawg commented:
I've also pushed the fix for the _async_send_response

RogerSelwyn commented:
> What improvements are you looking for, and in particular what do you think is missing? Everything seems to work fine which is strange to me, I just looked at the storage files the localcalendarstore creates and the serialized events don't seem to have as much details as I was expecting, but it works so... for example here is a serialized event I have in the storage file (I haven't removed anything for privacy, this is all it shows)

That doesn't seem to be storing start/end/etc

RogerSelwyn commented:
> It's not using the encoder your provided, which is probably why the dates aren't there because they don't encode. Not sure this is actually using the storage. Yes it saves stuff and pulls it, but not sure what it is using.

It was using the encode, didn't realise there is a two minute delay on it. However, because I hadn't supplied the beautify_key method, you'd 'tweaked' it :-) And it was therefore missing most of the keys. Try with my latest commit and see how you go. Maybe that will fix some other issues.

I'll try to look at some other bits in your comments into next week. My son's 30th tomorrow, so out for the day.

VanWardawg commented:
I did know I had removed that but I didn't think it was that important, my bad! Thanks for fixing that, I'll check it out myself and see how it looks, let me know what else might be missing or that you need help on before we check this in :)

RogerSelwyn commented:
> I did know I had removed that but I didn't think it was that important, my bad! Thanks for fixing that, I'll check it out myself and see how it looks, let me know what else might be missing or that you need help on before we check this in :)

What concerns me is that it was working when clearly the storage wasn't storing the data. So I'm not sure how the storage is helping, is it all in memory? I'll work through it because I think it is worthwhile, but may take a while.

VanWardawg commented:
It might be for the most part right now, there were some things I removed from the Gcal implementation (in regards to the timeline code) and it might be that normally it only loads X amount in memory but I might have removed that limitation...
It works though haha

VanWardawg commented:
gcal also uses pagination to slowly import/update the data

VanWardawg commented:
but even the sorcery ai comment shows it should work so maybe it was working anyways... not sure

VanWardawg commented:
And the beautify key code works great, all the data I'd expect is there now.

RogerSelwyn commented:
So you are aware, I'm breaking the elements your gathered gcal_sync back into a similar structure as in gcal. This really only so I can understand what is going on, I'll probably gather them back together at some point. This makes coordinator.py much simpler.

So if you see any commits over the next days, don't panic. Day out tomorrow, so I doubt much will get done.

It looks like for google calendar, it only uses one storage file. Is there a reason you broke it out into multiple for MS365?

VanWardawg commented:
No problem, at this point I'm fine with handing this over to you to change, update and adapt as necessary to fit your needs as you will be maintaining this longer term than I. I just know I wanted to have a quicker calendar (speed wise) solution for my needs and this has achieved that for me and I wanted to try to bring it in some form to the official releases. Hopefully this helped and didn't just create more work for you.

No reason, if gcal did one file and I did multiple it was just my bad, having one file should be fine as its uses a "scope" to only pull items from the file for the specific calendar. Sounds like a better approach than the one I was doing.

RogerSelwyn commented:
I have to have a serious think on this one, because the integration no longer fetches data for time periods outside the -60/+90. You can check this by paging forward on the calendar, which should load events as you page. get_events in MS365CalendarSyncCoordinator says it will manually request them, but it seems to just try to get them from memory, and if they aren't there, then nothing is returned.

As far as I can see it loads everything into memory (potentially from the store) at the start, then saves to store whenever it updates, but never retrieves anything from store after initial load. 

I may be reading it wrong of course, I plan to spend time tomorrow to see if I can get it to work.

VanWardawg commented:
I've fixed the missing data outside of the time period (I was just calling the function on the wrong class), as for having it in memory, it does keep it in memory but it is also loading it from the store successfully, my most recent change adds some more debug log lines that will show its calling the expected functions:
Syncing Calendar Events: AQMkADAwATMwMAItYjYyOS0zNTc5LTAwAi0wMAoARgAAAwFauBQmPy1Eif70yCxgWX8HAPQ6RsMlfYhHrAYaya5uDJYAAAIBBgAAAPQ6RsMlfYhHrAYaya5uDJYAB8RDhA4AAAA=
Saving data to store
Updating Data from API Endpoint AQMkADAwATMwMAItYjYyOS0zNTc5LTAwAi0wMAoARgAAAwFauBQmPy1Eif70yCxgWX8HAPQ6RsMlfYhHrAYaya5uDJYAAAIBBgAAAPQ6RsMlfYhHrAYaya5uDJYAB8RDhA4AAAA=
Loading data from store
Created timeline of 33 events -- Thisi is where we are casting the json objects to Event objects


I'll push my fix, please let me know if that works better. I've also simplified it to a single storage file.

RogerSelwyn commented:
Thanks, this helps. I had wondered if that was the problem, just hadn't got to the point of getting into detail. I'd noticed that one of the sync methods wasn't referenced anywhere.

Whilst it looks like it is retrieving data from disk, if you follow through that process, the async_load in the LocalCalendarStore looks like this:

```python
    async def async_load(self) -> dict[str, Any] | None:
        """Load data."""
        _LOGGER.debug("Load data")
        if self._data is None:
            _LOGGER.debug("Load from store")
            self._data = await self._store.async_load() or {}
        return self._data
```
As far as I can see `self._data = await self._store.async_load() or {}` will only get called once at startup, then self._data is populated and held in memory.

VanWardawg commented:
for the store loading it does appear like you are correct, however this is the exact behavior of the gcal sync. The main difference is the fetching of the data is different (since it does pagination) but it looks like it holds it in memory as well. Not sure if this isn't what you wanted but it appears like that is at least what the google calendar integration is doing.

RogerSelwyn commented:
That's how I felt about it as well, but to some extent I wanted to validate that is what you saw. I pretty much wonder why it gets stored to disk, when it appears never to be used from there apart from startup (which has minor benefit). Seems an awful lot of faffing around with storage to not use it.

RogerSelwyn commented:
Anyways, I'm working through fixing the tests to match.

RogerSelwyn commented:
Just one previous test to fix, nothing really to do with the new methods. Also a few more bits needed to provide coverage. Also removed a few items that aren't really needed.

VanWardawg commented:
great thanks for that!

RogerSelwyn commented:
@VanWardawg you will have seen that I have made a lot of changes to the code, much of which pulls it towards the same standard as the other integrations, which in essence makes it easier for me to remember how it works when I'm doing maintenance across all 4 integrations. I'd don't believe any of the changes will make it perform any differently to your input.

VanWardawg commented:
I totally understand, this is your project and I'm glad you can adopt it into what works best for the long term!

RogerSelwyn commented:
> "Can you expand on the issues you had?" When new data gets pulled in the merge I was doing, it was not removing the deleted items (from other devices) that were no longer returned from the API. This wasn't bad as I have it syncing the whole time range and do manual queries for any api call outside of the range but if/when this sync wants to slowly add/sync information it would need to be able to merge/add items while removing deleted items.
I had a look at adding pagination to the data pull, but I'm not sure it can be done. The library is not async at all, so doesn't return an async iter to work over. It does use batching, and will pull data in those batches, but in essence because the integration processes all events (e.g. for filtering) before storing and displaying I don't think it makes any difference. 

So I'm not sure slow sync will happen, but it is the sort of thing that will nag at me. The MS Graph Api does support getting deltas for endpoints, but the library doesn't support that at all, but it might be something I can look at getting added.

> "Is it the range of data to pull? Or how long to store for?: Yes, right now I have the min/max time to sync hardcoded in the coordinator but my thoughts would be having this as a configureable setting Same goes for the data sync frequency, I have it configured for every 7 minutes but if people aren't changing or don't mind slowly sync times between devices (for less powered devices, etc.) then they could adjust it.
 
> Current values I mean for this are: SYNC_EVENT_MIN_TIME = timedelta(days=-60) SYNC_EVENT_MAX_TIME = timedelta(days=90) and MIN_TIME_BETWEEN_UPDATES = timedelta(minutes=7)

I've added a configuration option for how often to sync, but I've left the -60/90 as is for now. I think people will confuse it with the existing config option on the devices. Maybe I should force them to use the get_events(start, end) service for retrieving there events, rather than putting them on the extra attributes.

> What improvements are you looking for, and in particular what do you think is missing? Everything seems to work fine which is strange to me, I just looked at the storage files the localcalendarstore creates and the serialized events don't seem to have as much details as I was expecting, but it works so... for example here is a serialized event I have in the storage file (I haven't removed anything for privacy, this is all it shows) { "AQMkADAwATMwMAItYjYyOS0zNTc5LTAwAi0wMAoARgAAAwFauBQmPy1Eif70yCxgWX8HAPQ6RsMlfYhHrAYaya5uDJYAB8Qu_PUAAAD0OkbDJX2IR6wGGsmubgyWAAfHSmVWAAAA": { "calendar_id": null, "object_id": "AQMkADAwATMwMAItYjYyOS0zNTc5LTAwAi0wMAoARgAAAwFauBQmPy1Eif70yCxgWX8HAPQ6RsMlfYhHrAYaya5uDJYAB8Qu_PUAAAD0OkbDJX2IR6wGGsmubgyWAAfHSmVWAAAA", "body_type": "html", "has_attachments": false, "ical_uid": null, "is_cancelled": false, "is_organizer": true, "locations": [], "online_meeting_url": null, "online_meeting": null, "series_master_id": null, "web_link": null },

We sorted this one.

RogerSelwyn commented:
In terms of release timeline, I'm away from Friday for a week, so I'd rather not release immediately before I'm away. It's really hard to debug (and I prefer not to) on my live systems, when I don't want to just hack the config around.

So probably the week after next.

VanWardawg commented:
Sounds good, and no issues, I've manually installed this so I have it running, it would just be nice to not have to manually override it for the next update, thanks!

VanWardawg commented:
Awesome, glad I could be of help!

RogerSelwyn commented:
In this version I have added some extra config variables for min/max time to sync. Currently it defaults to how it would behave now i.e. it syncs at minimum what is in calendars.yaml, every 30 seconds. I might change this default to something longer, but what I want to enable is for people to have the flexibility to have it as now, as well as how you want it.

For myself, I never really look at the calendars pane (or a similar view), so don't really care about performance of that. What I do want is that it updates quickly when I modify an event.

RogerSelwyn commented:
fyi - I've just started testing this in my production environment. Should release as a beta tomorrow/Tuesday

VanWardawg commented:
Awesome thanks! I look forward to installing the new supported version!
